float32 roll
float32 upper
float32 lower